### PR TITLE
fix: Handle conflicting changelogs in concurrent Merkle tree

### PR DESCRIPTION
**Pull request by vadorovsky**
Restore the usage of `ConcurrentMerkleTreeError::CannotUpdateLeaf` error. Throwing an error when any changelog entry that modifies the same leaf we try to update is a right thing to do.

In case of indexed Merkle tree, we can ignore this error. We are already patching the leaf using the indexed changelog.

Other changes:

* Fix `test_100_nullify_mt`:
  * Each Merkle tree update was pointing to old `changelog_index`. And the crank Merkle tree was not being updated after each operation.
* Implement `Clone` in `BoundedVec`, to make cloning of Merkle proofs possible.
* Add more tests:
  * Check whether elements returned in `ConcurrentMerkleTree::changelog_entries`
    are as expected.
  * Test proving that we need to `skip(1)` during Merkle proof update.
  * Test with an update on wrapped-around changelog.

### CODE REVIEW

File: merkle-tree/concurrent/src/changelog.rs [hunk 1]
```diff
@@ -62,21 +60,10 @@ impl<const HEIGHT: usize> ChangelogEntry<HEIGHT> {
...
             proof[intersection_index] = self.path[intersection_index];
-        } else {
-            // This case means that the leaf we are trying to update was
-            // already updated. Therefore, updating the proof is impossible.
-            // We need to return an error and request the caller
-            // to retry the update with a new proof.
-            //
-            // TODO(vadorovsky): Re-visit optional throwing of this error.
-            // if !allow_updates_in_changelog {
-            //     return Err(ConcurrentMerkleTreeError::CannotUpdateLeaf);
```
Comment (ananas-block) [inline merkle-tree/concurrent/src/changelog.rs:78]: As we established we still need this error because otherwise we undo previous update.
However, for the address Merkle tree update we should patch proof and leaf outside of the update function and update the changelog index accordingly.
We can do that since we have changelog for the latest indexed array changes of the address Merkle tree.

Comment (vadorovsky) [reply merkle-tree/concurrent/src/changelog.rs:77]: Restored the error

File: merkle-tree/concurrent/src/lib.rs [hunk 2]
```diff
@@ -792,7 +792,12 @@ where
         leaf_index: usize,
         proof: &mut BoundedVec<[u8; 32]>,
     ) -> Result<(), ConcurrentMerkleTreeError> {
-        for changelog_entry in self.changelog.iter_from(changelog_index) {
+        // Iterate over changelog entries starting from the requested
+        // `changelog_index`.
+        //
+        // Since we are interested only in subsequent, new changelog entries,
+        // skip the first result.
+        for changelog_entry in self.changelog.iter_from(changelog_index).skip(1) {
```
Comment (vadorovsky) [inline merkle-tree/concurrent/src/lib.rs:800]: For more context - I was getting `CannotUpdateLeaf` errors in `test_100_nullify_mt` before this change. If you remove the `skip(1)`, that test is going to break.

Why? The thing is that we need to only process changelog entries **newer** than the given one. Before applying this change, I was running into edge cases, where:

* A leaf was appended, the changelog index saved.
* Some other operations were done.
* The previously appended leaf was being nullified, the saved changelog index (coming from the insertion) was being used.
* The iteration over changelog started from the insertion event. `CannotUpdateLeaf` was thrown.

You might be wondering - what if we just did `self.changelog.iter_from(changelog_index + 1)`? Wouldn't that be cleaner?

Unfortunately, that wouldn't work in case when `changelog_index` is up to date. Pointing to up-to-date `changelog_index + 1` would result in pointing to the oldest element in the whole CyclicBouldedVec, and therefore, in the full iteration over all elements.

An another alternative I was considering was skipping the first element by default in `iter_from`, but that feels wrong if we think about CyclicBoundedVec as a self-contained thing.

Comment (ananas-block) [reply merkle-tree/concurrent/src/lib.rs:800]: True, provided the change_log index the caller provides is in sync with the root of the Merkle proof we are not interested in the changelog that was inserted when that root was created because our Merkle proof already includes that state.
This means that we have been doing unnecessary computation all the time right?
On the other hand we also cannot send changelog index + 1 to the chain because this is going to incur an iteration over the complete changelog array in case that changelog_index is the most recent changelog index  as you describe above.

Thus, skipping the first element sounds like a good solution to me.

File: merkle-tree/indexed/tests/tests.rs [hunk 3]
```diff
@@ -595,40 +598,42 @@ pub fn functional_non_inclusion_test() {
...
+//
+//     onchain_indexed_merkle_tree
+//         .update(
+//             change_log_index.clone(),
+//             address_bundle_1.new_element.clone(),
+//             old_low_address_1.clone(),
+//             old_low_address_next_value_1,
+//             &mut low_element_proof_1,
+//         )
+//         .unwrap();
```
Comment (ananas-block) [inline merkle-tree/indexed/tests/tests.rs:627]: leftover?

Comment (vadorovsky) [reply merkle-tree/indexed/tests/tests.rs:627]: Not really. I know it's commented code, but we want to uncomment it once concurrency is fixed. I adjusted this code to my changes.

Comment (vadorovsky) [reply merkle-tree/indexed/tests/tests.rs]: Actually, I kicked it out now. I will try to fix the address MT concurrency in a separate PR.

File: merkle-tree/concurrent/tests/tests.rs [hunk 4]
```diff
@@ -2204,3 +2220,698 @@ fn test_subtree_updates() {
...
+            println!("{i}: {:?}", changelog_entry.path[0]);
+        }
+
+        let changelog_entries = merkle_tree
+            .changelog_entries(merkle_tree.changelog.first_index())
+            .collect::<Vec<_>>();
+        assert_eq!(changelog_entries.len(), leaves.len() - 1);
+
+        for (leaf, changelog_entry) in leaves.iter().skip(1).zip(changelog_entries) {
+            assert_eq!(&changelog_entry.path[0], leaf);
```
Comment (ananas-block) [inline merkle-tree/concurrent/tests/tests.rs:2681]: can we assert the complete path here?

Comment (ananas-block) [reply merkle-tree/concurrent/tests/tests.rs:2681]: can we also assert the other fields of the changelog object?

File: merkle-tree/concurrent/tests/tests.rs [hunk 5]
```diff
@@ -2204,3 +2220,698 @@ fn test_subtree_updates() {
...
+
+    // Try to update the original `leaf` with an outdated proof and changelog
+    // index. Expect an error.
+    let new_leaf: [u8; 32] = Fr::rand(&mut rng)
+        .into_bigint()
+        .to_bytes_be()
+        .try_into()
+        .unwrap();
+
+    let res = merkle_tree.update(changelog_index, &leaf, &new_leaf, 0, &mut proof);
```
Comment (ananas-block) [inline merkle-tree/concurrent/tests/tests.rs:2885]: can we add the case that still works, followed by this case to demonstrate that it works when expected and throw exactly when expected?

Comment (vadorovsky) [reply merkle-tree/concurrent/tests/tests.rs:2888]: Done

File: merkle-tree/concurrent/tests/tests.rs (hunk 4 above)
Comment (vadorovsky) [reply merkle-tree/concurrent/tests/tests.rs]: Done

File: merkle-tree/concurrent/tests/tests.rs (hunk 5 above)
Comment (ananas-block) [reply merkle-tree/concurrent/tests/tests.rs:2937]: ty!

**Review by ananas-block — APPROVED** [review]
